tss-esapi/src/utils/credential.rs

```diff
@@ -0,0 +1,287 @@
+
+        // B.10.4 RSA Secret Sharing for Credentials
+        // https://trustedcomputinggroup.org/wp-content/uploads/TPM-2.0-1.83-Part-1-Architecture.pdf#page=302
+        let random_seed = {
+            let mut out = Key::<TpmHmac<EkHash>>::default();
+            rng.try_fill(out.as_mut_slice()).map_err(|e| {
+                error!("RNG error: {e}");
+                Error::local_error(WrapperErrorKind::InternalError)
+            })?;
+            out
+        };
+
+        // The random seed is then encrypted with RSA-OAEP
+        //
+        // B.4 RSAES_OAEP
+        // https://trustedcomputinggroup.org/wp-content/uploads/TPM-2.0-1.83-Part-1-Architecture.pdf#page=297
+        //
+        // The label is a byte-stream whose last byte must be zero
+        //
+        // B.10.4. RSA Secret Sharing for Credentials
+        // https://trustedcomputinggroup.org/wp-content/uploads/TPM-2.0-1.83-Part-1-Architecture.pdf#page=302
+        //
+        // The label is going to be "IDENTITY" for secret sharing.
+        let encrypted_seed = {
+            let padding = Oaep::new_with_label::<EkHash, _>(b"IDENTITY\0".to_vec());
+            let enc_data = ek_public
+                .encrypt(&mut rng, padding, &random_seed[..])
+                .map_err(|e| {
+                    error!("RSA OAEP encryption error: {e}");
+                    Error::local_error(WrapperErrorKind::InternalError)
+                })?;
+            enc_data
+        };
+        let encrypted_secret = EncryptedSecret::from_bytes(&encrypted_seed)?;
+
+        match secret_to_credential::<EkHash, EkCipher>(random_seed, secret, &key_name)? {
+            Ok(id_object) => return Ok((id_object, encrypted_secret)),
+            Err(WeakKeyError) => {
+                // 11.4.10.4 Rejection of weak keys
+                // https://trustedcomputinggroup.org/wp-content/uploads/TPM-2.0-1.83-Part-1-Architecture.pdf#page=82
+
+                // The Key was considered weak, and we should re-run the creation of the encrypted
+                // secret.
+                continue;
+            }
+        }
+    }
+}
+
+fn secret_to_credential<EkHash, EkCipher>(
+    seed: Key<TpmHmac<EkHash>>,
+    secret: &[u8],
+    key_name: &Name,
+) -> Result<WeakResult<IdObject>>
+where
+    EkHash: Digest + EagerHash + FixedOutputReset,
+    <EkHash as OutputSizeUser>::OutputSize: Mul<U8>,
+    <<EkHash as OutputSizeUser>::OutputSize as Mul<U8>>::Output: Unsigned,
+    <<EkHash as EagerHash>::Core as OutputSizeUser>::OutputSize: ArraySize + Mul<U8>,
+    <<<EkHash as EagerHash>::Core as OutputSizeUser>::OutputSize as Mul<U8>>::Output: Unsigned,
+
+    EkCipher: KeySizeUser + BlockCipherEncrypt + KeyInit,
+    <EkCipher as KeySizeUser>::KeySize: Mul<U8>,
+    <<EkCipher as KeySizeUser>::KeySize as Mul<U8>>::Output: ArraySize,
+{
+    // Prepare the sensitive data
+    // this will be then encrypted using AES-CFB (size of the symmetric key depends on the EK).
+    let mut sensitive_data = {
```

Comment:
Should this be zeroized once we're done with it?

Reply:
it's rewritten in-place

Reply:
added a comment to make it clear

Reply:
Actually no, you're right, if an error occurs then it's not zeroized anymore.